### DISCUSSION TITLE
Display normals for point cloud and colour each point based on a third value

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#510] [Q&A] [answered]
Hi!
I want to display the normals for a point cloud and modify the colours of the points based on a third value (the mean curvature), and add a colourbar. I have the points, normals and curvature already, just need some instruction on how to put it all together. If you could give an example, that would be great also!

Thanks 
J

**Comment by marcomusy** [comment]
E.g.:
```python
import vedo
import numpy as np

pts = vedo.Points(np.random.randn(1000000,3))
pts.pointdata['Curvature'] = pts.points() # whatever, i'm making this up!

# add an array which corresponds to the z component
pts.pointdata['ZCurvature'] = pts.pointdata['Curvature'][:,2] # numpy array

pts.cmap('jet', "ZCurvature").addScalarBar()

vedo.show(pts, axes=1)
```
![Screenshot from 2021-11-05 16-45-31](https://user-images.githubusercontent.com/32848391/140538749-bfe636dd-fb90-452b-bcc1-2679ff364983.png)

**Reply by jkissi** [reply]
Thanks for your response. Your example helped me alot and made me realise that my curvature estimation function **_isn't working_**. Hahaha, so thanks for your help there!

![curvature_edit](https://user-images.githubusercontent.com/8246894/140575389-e9e4ef14-ce92-4669-8807-0271b3b9836f.png)

Was wondering if there's a way to display a numpy array of point normal values? I tried using the NormalLines() function for meshes to generate normals and display them:

```
    vpts1 = ve.Points(points, r=3)
    vpts2 = ve.Mesh(points)
    norms = ve.NormalLines(vpts2, atCells=False)
    vpts1.pointdata['Curvature'] = avg_curv
    vpts1.cmap('jet', "Curvature").addScalarBar()
    ve.show(vpts1, norms, axes=1).close()
```

and I got this:

![sidewaysnormals_edit](https://user-images.githubusercontent.com/8246894/140580023-17bef004-6344-47d0-95d2-282e15439b19.png)

You can see that all the normals are sideways. Any idea why this could be happening? If you can suggest a way to add and display them manually, I can add pre-calculated normals that I know are correct. As always, cheers in advance for you help.

Cheers 
j

**Comment by marcomusy** [comment, marked as answer]
HI, just use your numpy vectors of normals and add them to the point positions, then use `Lines()`.
let me know if you need help,
marco

**Reply by jkissi** [reply]
Hi
Think I managed to get it down:
![goodnormals_edit](https://user-images.githubusercontent.com/8246894/140836703-91a2949f-9ba7-4518-976a-756d0eeb59a7.png)

Here's the code:

```
    vp = ve.Plotter(bg='black', sharecam=1)

    # generate to the Points obj
    vpts = ve.Points(points,  r=3)

    # generate normal lines
    norms = ve.Lines(vpts, normals, alpha=0.25, scale=0.02)

    vp.show(vpts, norms, axes=1).close()
```
I think It's pretty efficient looking.... but as always, any suggestions welcome.

Cheers!!
J

**Reply by marcomusy** [reply]
looks soft and fluffy i love it :)

so what you call normals should be vpts + normals i guess..

one thing you can do is to use vmin and vmax in cmap() to adjust the range of the scalar visualization if needed..
